wandb/sdk/artifacts/artifact_file_cache.py

```diff
@@ -38,12 +38,19 @@ def __init__(self, cache_dir: StrPath) -> None:
 
         # NamedTemporaryFile sets the file mode to 600 [1], we reset to the default.
         # [1] https://stackoverflow.com/questions/10541760/can-i-set-the-umask-for-tempfile-namedtemporaryfile-in-python
-        umask_cmd = (sys.executable, "-c", "import os; print(os.umask(22))")
-        umask = int(subprocess.check_output(umask_cmd))
-        self._sys_umask = umask
+        self._sys_umask = self._get_sys_umask_threadsafe()
 
         self._override_cache_path: Optional[StrPath] = None
 
+    @staticmethod
+    def _get_sys_umask_threadsafe() -> int:
+        # Workaround to get the current system umask, since
+        # - `os.umask()` isn't thread-safe
+        # - we don't want to inadvertently change the umask of the current process
+        # See: https://stackoverflow.com/questions/53227072/reading-umask-thread-safe
+        umask_cmd = (sys.executable, "-c", "import os; print(os.umask(22))")
+        return int(subprocess.check_output(umask_cmd))
```

Comment:
I know this was in the original code, but I don't understand why we're reading the umask at all. Why not use this in `_cache_opener` instead?

```python
os.chmod(temp_file.name, 0o600)
```

(read+write for owner, no permissions for anyone else)

Reply:
> Why not use this in `_cache_opener` instead?
> 
> ```python
> os.chmod(temp_file.name, 0o600)
> ```
> 
> (read+write for owner, no permissions for anyone else)

I may still be missing context for the orig code (cc @ibindlish in case I am) -- but from what I gather / infer so far:

- `temp_file` is already created with mode `0o600` (turns out this is intentionally [hard-coded](https://github.com/python/cpython/blob/1dc9a4f6b20148fd4ef2eb2800a6c65224828181/Lib/tempfile.py#L256) in the `tempfile` module)
- This would cause the final destination file(s) (e.g. what the user passes to `Artifact.download(...)`) to have mode `0o600` as well.  We could hard-code some other permission for the destination file, but AFAICT using the umask avoids that.  We could also keep the final file as `0o600`, but I suspect this would break some users' workflows -- or at least be significantly more inconvenient :\

Reply:
So this PR originally did the umask read in the cache_opener: https://github.com/wandb/wandb/pull/6437/files#diff-e9653d64fe6417b163a37e18de96fc53ecf0a3dc3e4820e6a604defda7a0d083

But that caused a performance regression for artifact downloads and was fixed here: https://github.com/wandb/wandb/pull/6535/files#diff-e9653d64fe6417b163a37e18de96fc53ecf0a3dc3e4820e6a604defda7a0d083

Reply:
> This would cause the final destination file(s) [...] to have mode 0o600 as well

I don't understand why this requires us to touch `umask`. Shouldn't we set the permissions explicitly on the resulting file?

The line `os.chmod(temp_file.name, 0o666 & ~self._sys_umask)` causes permissions on the tempfile to be a subset of `0o666` (read-write for all), so using `0o666` would be no more restrictive and shouldn't break workflows. Basically I don't see why we need to respect the `umask` here at all.

I read through the original issues, but it seems that what they're looking for is for us to respect the _original file's permissions_, not the process's `umask`. So if `dataset.h5` has permissions set to 002, then after using

```python
run.log_artifact(data="./dataset.h5", name="my_data", type="dataset")
...
artifact = run.use_artifact("my_data:latest")
artifact.download()
```

the downloaded file should have the same permissions. Is my understanding correct?

---

Anyway, since this PR doesn't change that code, we can revisit this later, but this `subprocess` invocation seems like a code smell to me. I'm also curious whether this issue happens with `wandb-core`---since the 0.18 release is coming up, the fix needs to work in core as well; CC @kptkin

Reply:
I guess in that case we'd have to save information about the logged file's permissions in the manifest entry maybe. haven't thought about it that way. In general though (outside W&B), I think when you download files, the downloaded file is expected to have your system's umask

We can't always assume that the person logging a file to an artifact will be the one downloading it too. So I'm not sure that keeping track of the original permission set is the right way to go here

Reply:
Not sure this'll be resolved in this PR, but I've created a ticket from the thread here for posterity: https://wandb.atlassian.net/browse/WB-19618